// Cloud Dispatcher Service - Fleet-wide RPC dispatch
// Provides gRPC API for invoking methods on vehicle services
syntax = "proto3";

package ifex.cloud.dispatcher;

import "dispatcher-rpc-envelope.proto";

// === RPC Status (extended for cloud) ===

enum CloudRpcStatus {
    CLOUD_RPC_UNKNOWN = 0;
    CLOUD_RPC_PENDING = 1;          // Request sent, waiting for response
    CLOUD_RPC_SUCCESS = 2;          // Response received successfully
    CLOUD_RPC_FAILED = 3;           // Vehicle returned error
    CLOUD_RPC_TIMEOUT = 4;          // No response within timeout
    CLOUD_RPC_VEHICLE_OFFLINE = 5;  // Vehicle not reachable
    CLOUD_RPC_SERVICE_UNAVAILABLE = 6;
    CLOUD_RPC_METHOD_NOT_FOUND = 7;
    CLOUD_RPC_INVALID_PARAMETERS = 8;
    CLOUD_RPC_TRANSPORT_ERROR = 9;  // Kafka/MQTT error
    CLOUD_RPC_CANCELLED = 10;       // Request cancelled
}

// === Call Method (Synchronous) ===

message CallMethodRequest {
    string vehicle_id = 1;
    string service_name = 2;
    string method_name = 3;
    string parameters_json = 4;        // JSON-encoded parameters
    uint32 timeout_ms = 5;             // Default: 30000 (30 seconds)
    string requester_id = 6;           // Who initiated this call (for tracking)
    map<string, string> metadata = 7;  // Optional metadata for tracing
}

message CallMethodResponse {
    string correlation_id = 1;
    CloudRpcStatus status = 2;
    string result_json = 3;            // JSON-encoded result
    string error_message = 4;
    uint32 duration_ms = 5;            // Round-trip time
    string service_endpoint = 6;       // Which endpoint handled the call
}

// === Call Method Async ===

message CallMethodAsyncRequest {
    string vehicle_id = 1;
    string service_name = 2;
    string method_name = 3;
    string parameters_json = 4;
    uint32 timeout_ms = 5;
    string requester_id = 6;
    map<string, string> metadata = 7;
}

message CallMethodAsyncResponse {
    string correlation_id = 1;         // Use this to poll for result
    bool accepted = 2;                 // Whether request was accepted
    string error_message = 3;          // Error if not accepted
}

// === Get RPC Status ===

message GetRpcStatusRequest {
    string correlation_id = 1;
}

message GetRpcStatusResponse {
    string correlation_id = 1;
    string vehicle_id = 2;
    string service_name = 3;
    string method_name = 4;
    CloudRpcStatus status = 5;
    string result_json = 6;
    string error_message = 7;
    uint32 duration_ms = 8;
    bool completed = 9;
    uint64 created_at_ms = 10;
    uint64 responded_at_ms = 11;
    string requester_id = 12;
}

// === List RPC Requests ===

message ListRpcRequestsRequest {
    string vehicle_id_filter = 1;      // Optional
    string service_name_filter = 2;    // Optional
    string requester_id_filter = 3;    // Optional
    bool pending_only = 4;             // Only in-flight requests
    bool failed_only = 5;              // Only failed requests
    uint64 since_ms = 6;               // Only requests after this timestamp
    int32 page_size = 10;
    string page_token = 11;
}

message RpcRequestSummary {
    string correlation_id = 1;
    string vehicle_id = 2;
    string service_name = 3;
    string method_name = 4;
    CloudRpcStatus status = 5;
    bool completed = 6;
    uint64 created_at_ms = 7;
    uint64 responded_at_ms = 8;
    uint32 duration_ms = 9;
    string requester_id = 10;
}

message ListRpcRequestsResponse {
    repeated RpcRequestSummary requests = 1;
    string next_page_token = 2;
    int32 total_count = 3;
}

// === Cancel RPC Request ===

message CancelRpcRequest {
    string correlation_id = 1;
}

message CancelRpcResponse {
    bool success = 1;
    string error_message = 2;
}

// === Call Fleet Method (invoke on multiple vehicles) ===

message CallFleetMethodRequest {
    repeated string vehicle_ids = 1;   // Target vehicles (explicit list)
    string fleet_id = 2;               // OR all vehicles in fleet
    string region = 3;                 // OR all vehicles in region
    string service_name = 4;
    string method_name = 5;
    string parameters_json = 6;
    uint32 timeout_ms = 7;
    string requester_id = 8;
    bool parallel = 9;                 // Execute in parallel (default: true)
}

message FleetCallResult {
    string vehicle_id = 1;
    string correlation_id = 2;
    CloudRpcStatus status = 3;
    string result_json = 4;
    string error_message = 5;
    uint32 duration_ms = 6;
}

message CallFleetMethodResponse {
    int32 total_vehicles = 1;
    int32 successful = 2;
    int32 failed = 3;
    int32 pending = 4;
    repeated FleetCallResult results = 5;
}

// === gRPC Service Definition ===

service CloudDispatcherService {
    // Synchronous RPC call (blocks until response or timeout)
    rpc CallMethod(CallMethodRequest) returns (CallMethodResponse);

    // Asynchronous RPC call (returns correlation_id for polling)
    rpc CallMethodAsync(CallMethodAsyncRequest) returns (CallMethodAsyncResponse);

    // Get status/result of an RPC request by correlation_id
    rpc GetRpcStatus(GetRpcStatusRequest) returns (GetRpcStatusResponse);

    // List RPC requests for monitoring
    rpc ListRpcRequests(ListRpcRequestsRequest) returns (ListRpcRequestsResponse);

    // Cancel a pending RPC request
    rpc CancelRpc(CancelRpcRequest) returns (CancelRpcResponse);

    // Invoke method on multiple vehicles
    rpc CallFleetMethod(CallFleetMethodRequest) returns (CallFleetMethodResponse);
}
